### ISSUE TITLE
Segfault in in libsystemd-shared, trigged through nspawn shutdown

### ISSUE THREAD

**Issue by mphilipps** [systemd/systemd#12073]
**systemd version the issue has been seen with**
> 241

**Used distribution**
 > Debian Stretch with systemd from backports

**Expected behaviour you didn't see**
 > Regular shutdown

**Unexpected behaviour you saw**
 > Segfault in  libsystemd-shared-241.so

**Steps to reproduce the problem**
Every time I shutdown one of the affected containers it triggers a segfault.

hi,
Since we updated from 239 to 241 we reliably trigger a segfault in libsystemd-shared-241.so.
The setup affected by this is a bit exotic and might be cause of triggering the bug, so let me explain: 
**System A (Host):**
We have one physical host that is running Debian stretch with systemd 232. That host has a nspawn container with Debian Stretch and systemd 241.
**System B (Container manager):**
This container runs with `capability all` and to make reaching out to the host easier with `--bind-ro /proc/1/ns:/host`. 
Within that container runs another template container. (System C)
System B has several extra ip addresses that are used for socket activation, so that whenever someone connects through those, System B reaches out to System A to create an lvm snapshot, mounts it and then after preparing it, launches it as another nspawn container. 

Ever since we updated from 239 to 241 when one of these snapshot containers shuts down it triggers a segfault in libsystemd-shared. 

Systemd service unit for the affected container:
```
[Unit]
Description=Usercontainer for %i

[Service]
EnvironmentFile=/etc/default/%i.env
ExecStartPre=/usr/bin/nsenter --mount=/host/mnt --cgroup=/host/cgroup --uts=/host/uts --ipc=/host/ipc -- /sbin/lvcreate -s -L 3G -n ${SNAPSHOT} /dev/vg2/lxc_template_config
ExecStartPre=/usr/bin/nsenter --mount=/host/mnt --cgroup=/host/cgroup --uts=/host/uts --ipc=/host/ipc -- /bin/mount -o nouuid /dev/mapper/vg2-${SNAPSHOT} /vmroots/container-master2.tld/vmroots/%i
ExecStartPre=/etc/scripts/prepare_container.sh "${USER}" "${USE}" "${IP}"

ExecStart=/usr/bin/systemd-nspawn --keep-unit --boot --link-journal=try-guest --network-bridge=br16 --machine ${INTERFACE} --directory /vmroots/%i 3

ExecStopPost=/usr/bin/nsenter --mount=/host/mnt --cgroup=/host/cgroup --uts=/host/uts --ipc=/host/ipc -- /bin/umount /vmroots/container-master2.tld/vmroots/%i
ExecStopPost=/usr/bin/nsenter --mount=/host/mnt --cgroup=/host/cgroup --uts=/host/uts --ipc=/host/ipc -- /sbin/lvremove -f "/dev/vg2/${SNAPSHOT}"
KillMode=mixed
Type=notify
RestartForceExitStatus=133
SuccessExitStatus=133

[Install]
WantedBy=multi-user.target
```
Backtrace, unfortunately without debug symbols. Apparently there is no package for them in Debian.
```
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
Core was generated by `/usr/bin/systemd-nspawn --keep-unit --boot --link-journal=try-guest --network-b'.
Program terminated with signal SIGSEGV, Segmentation fault.
#0  0x00007fe4d02e710b in internal_hashmap_clear () from /lib/systemd/libsystemd-shared-241.so
(gdb) bt
#0  0x00007fe4d02e710b in internal_hashmap_clear () from /lib/systemd/libsystemd-shared-241.so
#1  0x00007fe4d02e720e in internal_hashmap_free () from /lib/systemd/libsystemd-shared-241.so
#2  0x00007fe4d0374c9c in fdset_free () from /lib/systemd/libsystemd-shared-241.so
#3  0x000055fd7cfba086 in ?? ()
#4  0x00007fe4d06d42e1 in __libc_start_main (main=0x55fd7cfb9ac0, argc=9, argv=0x7ffd70b11c88, init=<optimized out>, fini=<optimized out>,
    rtld_fini=<optimized out>, stack_end=0x7ffd70b11c78) at ../csu/libc-start.c:291
#5  0x000055fd7cfbc5ba in ?? ()
```

**Comment by poettering** [comment]
Please provide the backtrace with debug symbols. Otherwise there's little we can do, the issue is not actionable without that.

Also, can you run this through valgrind?

**Comment by mphilipps** [comment]
It has been a while that I used valgrind, is there anything specific you would want from valgrind?

**Comment by poettering** [comment]
running it with default args should be good enough.

**Comment by mphilipps** [comment]
Found the debug symbols:
```
[New LWP 11719]
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
Core was generated by `/usr/bin/systemd-nspawn --keep-unit --boot --link-journal=try-guest --network-b'.
Program terminated with signal SIGSEGV, Segmentation fault.
#0  0x00007fe4d02e710b in internal_hashmap_clear (h=h@entry=0x55fd7d843318, default_free_key=0x0, default_free_value=0x0) at ../src/basic/hashmap.c:881
881             free_key = h->hash_ops->free_key ?: default_free_key;
(gdb) bt
#0  0x00007fe4d02e710b in internal_hashmap_clear (h=h@entry=0x55fd7d843318, default_free_key=0x0, default_free_value=0x0) at ../src/basic/hashmap.c:881
#1  0x00007fe4d02e720e in internal_hashmap_free (h=0x55fd7d843318, default_free_key=<optimized out>, default_free_value=<optimized out>)
    at ../src/basic/hashmap.c:869
#2  0x00007fe4d0374c9c in set_free () at ../src/basic/set.h:12
#3  fdset_free (s=0x55fd7d843318) at ../src/shared/fdset.c:70
#4  0x000055fd7cfba086 in fdset_freep () at ../src/shared/fdset.h:39
#5  main (argc=<optimized out>, argv=<optimized out>) at ../src/nspawn/nspawn.c:4177

```

**Comment by mphilipps** [comment]
And valgrind:

```
Container con-51 has been shut down.                                  
--14783-- WARNING: unhandled amd64-linux syscall: 326                   
--14783-- You may be able to write your own handler.                               
--14783-- Read the file README_MISSING_SYSCALL_OR_IOCTL.      
--14783-- Nevertheless we consider this a bug.  Please report     
--14783-- it at http://valgrind.org/support/bug_reports.html.
==14783== Invalid read of size 8                                  
==14783==    at 0x529810B: internal_hashmap_clear (hashmap.c:881)  
==14783==    by 0x529820D: internal_hashmap_free (hashmap.c:869)                                    
==14783==    by 0x5325C9B: set_free (set.h:12)             
==14783==    by 0x5325C9B: fdset_free (fdset.c:70)                            
==14783==    by 0x122085: fdset_freep (fdset.h:39)                                                  
==14783==    by 0x122085: main (nspawn.c:4177)   
==14783==  Address 0x10 is not stack'd, malloc'd or (recently) free'd 
==14783==                                                  
==14783== Process terminating with default action of signal 11 (SIGSEGV)
==14783==  Access not within mapped region at address 0x10 
==14783==    at 0x529810B: internal_hashmap_clear (hashmap.c:881)     
==14783==    by 0x529820D: internal_hashmap_free (hashmap.c:869)                                       
==14783==    by 0x5325C9B: set_free (set.h:12)                                                   
==14783==    by 0x5325C9B: fdset_free (fdset.c:70)                                                   
==14783==    by 0x122085: fdset_freep (fdset.h:39)
==14783==    by 0x122085: main (nspawn.c:4177)
==14783==  If you believe this happened as a result of a stack
==14783==  overflow in your program's main thread (unlikely but
==14783==  possible), you can try to increase the size of the 
==14783==  main thread stack using the --main-stacksize= flag.
==14783==  The main thread stack size used in this run was 8388608.
==14783==                 
==14783== HEAP SUMMARY:
==14783==     in use at exit: 8,292 bytes in 3 blocks
==14783==   total heap usage: 291 allocs, 288 frees, 2,201,645 bytes allocated
==14783==                 
==14783== LEAK SUMMARY:
==14783==    definitely lost: 0 bytes in 0 blocks
==14783==    indirectly lost: 0 bytes in 0 blocks
==14783==      possibly lost: 0 bytes in 0 blocks
==14783==    still reachable: 8,292 bytes in 3 blocks
==14783==         suppressed: 0 bytes in 0 blocks
==14783== Rerun with --leak-check=full to see details of leaked memory
==14783==                 
==14783== For counts of detected and suppressed errors, rerun with: -v
==14783== ERROR SUMMARY: 1 errors from 1 contexts (suppressed: 0 from 0)
```

**Comment by poettering** [comment]
thanks, found the issue! patch coming shortly

**Comment by poettering** [comment]
Fix waiting in #12078
